tools/flowc/backends/nim/fi2nim.flow

```diff
@@ -4,18 +4,76 @@ import tools/flowc/backends/nim/fi2nim_defines;
 import tools/flowc/backends/nim/fi2nim_utils;
 import tools/flowc/backends/nim/fi2nim_compile;
 import tools/flowc/manipulation/flatten_unions;
+import tools/flowc/manipulation/specialization;
+import sys/process;
 
 export {
 	fi2nim(program : FiProgram, cfg : FiNimConfig, callback : (int) -> void) -> void;
 }
 
-fi2nim(program0 : FiProgram, cfg : FiNimConfig, callback : (int) -> void) -> void {
+fi2nim(program0 : FiProgram, cfg0 : FiNimConfig, callback : (int) -> void) -> void {
+	cfg = if (cfg0.outputfile != "1") cfg0 else {
+		FiNimConfig(cfg0 with outputfile = changeFileExt(cfg0.flowfile, ".nim"));
+	}
 	fcVerbose(cfg, 0, "Making NIM '" + cfg.outputfile + "'...");
 
 	// Unions shouldn't contain other unions - only structs
-	// fiCompleteStructTypars
-	// fiRemoveImplicitStructTypars
-	program = program0 |> fiCompleteStructTypars |> fiFlattenUnions;
+	program = program0 |> fiRemoveImplicitStructTypars |> fiFlattenUnions;
+
+	is_explicit_polymorphic = \nm -> switch (lookupTree(program.names.structs, nm)) {
+		Some(struct): length(struct.typars) > 0;
+		None(): switch (lookupTree(program.names.unions, nm)) {
+			Some(union): length(union.typeparameters) > 0;
+			None(): true;
+		}
+	}
+
+	polymorphics = filterTree(collect_polymorphic_entities(program, false), \nm, __ -> 
+		// leave only true (=explicit) polymorhpic structs/unions and remove implicitly polymorphic
+		is_explicit_polymorphic(nm)
+	);
+	//println("polymorphics:\n" + strIndent(
+	//	superglue(tree2pairs(polymorphics), \p -> p.first + " => " + prettyFiType(dummyPretty, p.second, makeSet()), "\n")
+	//));
+	specializations = filterTree(
+		mapTree(
+			find_specializations(program, polymorphics, makeTree(), false), 
+			\specs -> mapSet(specs, \spec -> fiType2NimType(program.names, spec))
+		),
+		// leave only true (=explicit) polymorhpic structs/unions and remove implicitly polymorphic
+		\nm,__ -> is_explicit_polymorphic(nm)
+	);
+	//println("specializations:\n" + strIndent(
+	//	superglue(
+	//		tree2pairs(specializations), 
+	//		\p -> p.first + " =>\n" + strIndent(superglue(set2array(p.second), \t -> prettyFiType(dummyPretty, t, makeSet()), "\n")), 
+	//		"\n"
+	//	))
+	//);
+	structs = fold(program.traversal, [], \acc, mname ->
```

Comment:
how huge program.traversal could be? making millions of arrays not good from performance perspective